doc/man/man1/swbisparse.1
...\" $Header: /usr/src/docbook-to-man/cmd/RCS/docbook-to-man.sh,v 1.3 1996/06/17 03:36:49 fld Exp $
...\"
...\"	transcript compatibility for postscript use.
...\"
...\"	synopsis:  .P! <file.ps>
...\"
.de P!
\\&.
.fl			\" force out current output buffer
\\!%PB
\\!/showpage{}def
...\" the following is from Ken Flowers -- it prevents dictionary overflows
\\!/tempdict 200 dict def tempdict begin
.fl			\" prolog
.sy cat \\$1\" bring in postscript file
...\" the following line matches the tempdict above
\\!end % tempdict %
\\!PE
\\!.
.sp \\$2u	\" move below the image
..
.de pF
.ie     \\*(f1 .ds f1 \\n(.f
.el .ie \\*(f2 .ds f2 \\n(.f
.el .ie \\*(f3 .ds f3 \\n(.f
.el .ie \\*(f4 .ds f4 \\n(.f
.el .tm ? font overflow
.ft \\$1
..
.de fP
.ie     !\\*(f4 \{\
.	ft \\*(f4
.	ds f4\"
'	br \}
.el .ie !\\*(f3 \{\
.	ft \\*(f3
.	ds f3\"
'	br \}
.el .ie !\\*(f2 \{\
.	ft \\*(f2
.	ds f2\"
'	br \}
.el .ie !\\*(f1 \{\
.	ft \\*(f1
.	ds f1\"
'	br \}
.el .tm ? font underflow
..
.ds f1\"
.ds f2\"
.ds f3\"
.ds f4\"
.ta 8n 16n 24n 32n 40n 48n 56n 64n 72n 
.TH "swbisparse" "1"
 
.hy 0 
.if n .na
.SH "NAME"
swbisparse \(em Read and parse POSIX\&.7\&.2 metadata\&.
.SH "SYNOPSIS"
.PP
.nf
\fBswbisparse\fP {--installed|-o,--info|-i,--index|-p,--psf} [-b] [-n] [-l \fIlevel\fP] [\fIfile\fP]
.fi
.PP
.nf
\fBswbisparse\fP {-d,--decode} [\fIpackage_file\fP]
.fi
.SH "DESCRIPTION"
.PP
\fBswbisparse\fP reads and parses POSIX\&.7\&.2 metadata files: INDEX, INFO and psf files\&.
.PP
\fBswbisparse\fP writes an implementation defined format to stdout\&.
The format indicates the keyword type, (O)bject,
(A)ttribute, or, (E)xtended, and, containment level\&.  Newlines in multi-line values
are translated to \&'\\n\&' and output on a single line\&. 
.PP
.br
.PP
When decoding a package file, \fIpackage_file\fP, 
\fBswbisparse\fP emits a format that includes the length (in
bytes) of the attribute value\&.  The format consists of ASCII data\&.
Binary data in the input will cause an error\&.
.PP
.PP
.nf
\f(CWFor Example, the following PSF file: 

 distribution
 layout_version 1\&.0
 product
 description "This is a
 multi line value"
 tag p1
 fileset
 tag p2
 file
 path /etc/passwd

is translated to:

 F00 PSF:
 O01 distribution
 A02  layout_version 1\&.0
 O02  product
 A03   description "This is a\\nmulti line value"
 A03   tag p1
 O03   fileset
 A04    tag p2
 O04    file
 A05     path /etc/passwd\fR
.fi
.PP
.SS "Hewlett-Packard SD-UX Compatibility"
.PP
This parser does not use nor require the \fBend\fP keyword to terminate objects as
is commonly the practice for PSF files written for Hewlett-Packard\&'s SD-UX equipped systems\&.  The \fBend\fP 
keyword can appear in PSF, INDEX and INFO files, however, it will merely be scanned and the
token discarded\&.  Another incompatibility is based on this parser\&'s termination of an object definition
by the first extended keyword\&.  This restriction may not
exist in the SD-UX grammar because the object is bracketed by the nesting of \fBend\fP keywords and
not by the existence of the first object or extended keyword\&.
.SH "OPTIONS"
.PP
\fB--info\fP\fB-o\fP     Parse an INFO file\&.
.PP
\fB--installed\fP     Parse the installed software catalog file INSTALLED\&.
.PP
\fB--psf\fP\fB-p\fP     Parse a PSF file\&.
.PP
\fB--index\fP\fB-i\fP     Parse an INDEX file\&.
.PP
\fB-b\fP     Beautify by indention and emit a file
.RS
semantically identical to
the input file\&.
.RE
.PP
\fB-n\fP     Emit format that is marked up with the 
.RS
value\&'s lengths in bytes\&. 
.RE
.PP
\fB-l\fP \fIlevel\fP     Set initial containment level
.RS
to \fIlevel\fP\&.
.RE
.SH "APPLICABLE STANDARDS"
.PP
Conforming to POSIX\&.7\&.2, however, A stand-alone parser 
utility by this (or any) name is not specified in ISO/IEC 15068-2:1999
.SH "SEE ALSO"
.PP
sw(5)
.SH "IDENTIFICATION"
.PP
 Author: Jim Lowe 
 Version: 0\&.481
 Last Updated 2006-02-15
 Copying: GNU Free Documentation License v1\&.1
.SH "BUGS"
.PP
This program is mainly a testing and development tool\&.
.PP
Requirements for inclusion (exclusion) of certain keywords in (from) the 
parsed output is not enforced\&.  Quoted trailing whitespace in the attribute
value is discarded\&.
.\" created by instant / docbook-to-man, Fri 02 Nov 2018, 20:39


NEWS
GNU swbis NEWS  Noteworthy changes. Jul 2023 (2023-07)

version 1.13.3 - Jim Lowe,  2023-07-01
    NEW FIXES:
      * Building, compilation
       - fix errors related to use of extern that
         show up with newer GCC compilers
       - Fix sysmacros.h issue again, use correct name

version 1.13.2 - Jim Lowe,  2020-02-28
    NEW FIXES:
      * Building, compilation
       -fix conditional inclusion for sys/sysmacros.h
        which was not being included up to now
	(swsupplib/configure was re-generated).

      * <swpackage>
	- Now add "state available" to fileset definition as
	the spec requires.


version 1.13.1 - Jim Lowe,  2018-11-03
    NEW FIXES:
      * <swpackage>
        - fixed coredump (segfault) when processing PSFs having
         an attribute value with a length larger than 99999 bytes
         (e.g. happens when translating some Linux kernel RPMs).

      * <swign, swinstall, swlist, et.al.>
        - reorder options in the invocation of GNU tar (to support
          latest version).


    NEW FEATURES:
      * <swpackage> - added support for translating Debian packages
          (deb format) that use xz compression internally.

      * <swpackage> - added option to exclude system directories from
           a package (for use when translating Deb format packages).


version 1.13 - Jim Lowe,  2016-05-14
    * Minor fixes, see ./ChangeLog

version 1.12 - Jim Lowe,  2014-08-02

   NEW FEATURES:

     * pax extended headers are now supported for name and file
       attributes too long for traditional tar.

     * The '-H pax' option of GNU tar is specified for all uses
       of the system tar utility.  A runtime check is made to support
       generic tar as much as possible.
     
     * Now multiple signatures are supported both in a package archive
       and in installed software.

     * 'swpackage' now has capability to resign, add and delete
        signatures of a previously signed package.

   NEW FIXES, among many others (See ./ChangeLog).

     * <all utilities except swpackage>
       A rare hang when accessing the installed software catalog is
       probably fixed.

     * <swverify>
       Now the package catalog is not decoded if the package signature
       fails.

     * <swverify>
       Fixed hang on non-Linux kernels.

     * <swverify>
       Added a new feature to verify prelink'ed binaries
       using prelink(8) on systems that use prelink.

     * <swverify>
       Added new reporting form (--report-format=FORM) that
       very similar to rpm's output format for verification.

     * <swlist, swremove,>
       FIXED operation with mksh or pdksh as the shell command

     * <swinstall>
       FIXED support for catalog at location outside installed 
       root dir.  The starting current working directory for control
       scripts is now SW_ROOT

     * <lxpsf> (rpm translation>
       Fixed support of RPMTAG_DEFAULTPREFIX.
     
     * <lxpsf> (deb translation)
       Added support for xz compressed archives.

     * <lxpsf> (rpm translation>
       Many corner cases in RPM translation are now fixed. Translation,
       installation, and cross verification have been tested on many
       hundreds of packages (on CentOS release 6.5)


version 1.11 - Jim Lowe,  2011-04-23
* <swpackage>: FIXED a race condition resulting in a signing failure
* <lxpsf>: FIXED segfault when translating some .deb file
* <lxpsf>: added support .lzma compressed data files in a .deb file

version 1.10 - Jim Lowe,  2011-04-15
* <swpackage>: FIXED the translation mode busted in the 1.9
release.

version 1.9 - Jim Lowe,  2011-01-20
* <swinstall>: FIXED the to-stdout mode 'swinstall @-'

version 1.8 - Jim Lowe,  2011-01-16

Fixed defects in the error code paths of all the utilities
that occur when file system permissions deny success to an
installed catalog operation.

Re-wrote (and fixed) the synct_* routines used for reading installed
catalog data with an in-band EOF.  These routines could fail due to
certain data contents.

version 1.7 - Jim Lowe,  2010-12-18
* <swinstall>: FIXED a broken speed optimization
* <swlist>: FIXED crashes due to use of unsupported features.
* <swinstall>: Improved compatibility with systems that have neither GNU tar
nor pax.
* FIXED several long standing bugs affecting all utilities except
swpackage.

version 1.6 - Jim Lowe,  2010-09-08
* All utilities fixed to operate on relative path target (such as the
  login landing directory of a remote target).
* <swlist>: added features. support -v mode more support for lising
attributes
* <swinstall>: FIXED swinstall preview mode for non-existent target directory
* <swinstall>: FIXED preview ''swinstall -p'' deleting catalog entry.
* <swinstall>: ADDED checks for existing files.  --replacefiles and --keepfiles
option added
* <swpackage>: ADDED implements the extended option -x follow_symlinks=true
* <swpackage>: FIXED uid/gid handling for case of `-o owner,' in a explicit
file definition in a PSF.  Now, finally, hopefully full arbitrary
specification of uids and unames now work

version 1.5 - Jim Lowe,  2010-03-15

* <swverify>: Now has a verification summary for installed software.

* the utilities now support an absolute path location for
the installed software catalog that is outside the target path.
This is specified by the file URL syntax: file:///absolute/path/name

* <swconfig>: swconfig is completed and enters Beta testing. 

* <swpackage>: Improvements to the ''exclude'' directive in a PSF file

* Many other fixes

version 1.4 - Jim Lowe,  2009-07-06

* Added swpackage --regfiles-only option

* Fixed specification of extended options on the command line.

* Added support for decrypting/encrypting via gpg command

* Added support for LZMA compression.

* Fixed <swpackage> Now software_definitions in a PSF
are properly merged.

* Fixed <lxpsf> Now the contents of ./conffiles in .deb format package
are used to set the 'is_volatile' attribute.

version 1.3 - Jim Lowe,  2009-03-11


swprogs/swbisparse.cxx
// swbisparse.cxx - A stand-alone parser for INFO, INDEX and PSF files.

/*
 Copyright (C) 2001 Jim Lowe

*/
/*
 COPYING TERMS AND CONDITIONS:

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 3, or (at your option)
 any later version.
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
*/
#include "swuser_config.h"
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <limits.h>
#include <string>
#include <typeinfo>
#include "stream_config.h"

extern "C" {
#include "swparse.h"
#include "usgetopt.h"
#include "ugetopt_help.h"
#include "swlib.h"
#include "swfork.h"
#include "swevents.h"
#include "swheader.h"
#include "swheaderline.h"
#include "swcommon.h"
#include "swgp.h"
#include "fmgetpass.h"
#include "swutilname.h"
#include "swfdio.h"
#include "swextopt.h"
}

#include "swmain.h"  /* this file should be included by the main program file */
#include "swparser.h"
#include "swdefinitionfile.h"
#include "swpsf.h"
#include "swpackagefile.h"
#include "swexstruct.h"
#include "swexdistribution.h"

#include "swexfileset.h"
#include "swexproduct.h"
#include "swexpsf.h"
#include "swexdistribution.h"
#include "swscollection.h"
#include "swinstalled.h"
#include "swinfo.h"
#include "swoption.h"

extern int swparse_atlevel;
extern char  swlex_filename[512];

static int usage (char * progname, struct option  lop[])
{
    fprintf (stderr, "A POSIX-7.2 Metadata parser and testing and development.\n");
    fprintf (stderr, "utility of the swbis project.\n");
    fprintf (stderr, "Copyright (C) 2003 Jim Lowe\n");
    fprintf (stderr, "This software may only be redistributed under terms of the GNU GPL.\n");
    fprintf (stderr, "\nUsage:   swbisparse [options] {-i|-o|-p} [psf_file]\n");
    fprintf (stderr, "Write annotated, single line output to stdout.\n");
    fprintf (stderr, "\n");

       static struct ugetopt_option_desc help_desc[] =
             {
               {"indent", "", "Beautify only, output is semantically unchanged."},
               {"index", "", "parse according to INDEX file syntax."},
               {"installed", "", "parse according to INSTALLED file syntax."},
               {"info", "", "parse according to INFO file syntax."},
               {"psf", "", "parse according to PSF file syntax."},
               {"psfi", "", "parse an include file (attribute value)\n"
	       "             Internal use only."},
               {"option", "", "parse according to extended option file syntax."},
               {"PSF", "", "same as lower case equivalent."},
               {"INFO", "", "same as lower case equivalent."},
               {"INDEX", "", "same as lower case equivalent."},
               {"INSTALLED", "", "same as lower case equivalent."},
    	       {"debug", "", "Be verbose when parsing."},
               {"level", "LEVEL", "Set level to LEVEL."},
               {"with-line-lengths", "", "Include attribute value lengths in output"},
               {"posix-ignores", "", "Exclude certain attributes from output according to the Std."},
               {"help", "", "Show this help"},
               {0, 0, 0}
             };

	ugetopt_print_help(stderr, progname, lop, help_desc);

    exit (1);
}

int main (int argc, char *argv[])
{
  int fd, c=0, ret;
  char  *progname;
  swDefinitionFile * swdef = NULL;
  int oForm = SWPARSE_FORM_MKUP;
  int do_debug = 0;
 
  swparse_atlevel=0;
 
 static struct option long_options[] =
             {
               {"indent", 0, 0, 'b'},
               {"index", 0, 0, 'i'},
               {"installed", 0, 0, 200},
               {"info", 0, 0, 'o'},
               {"psf", 0, 0, 'p'},
               {"psfi", 0, 0, 'h'},
               {"option", 0, 0, 'd'},
               {"PSF", 0, 0, 'p'},
               {"INFO", 0, 0, 'o'},
               {"INDEX", 0, 0, 'i'},
               {"INSTALLED", 0, 0, 200},
               {"debug", 0, 0, 'v'},
               {"level", 1, 0, 'l'},
               {"with-line-lengths", 0, 0, 'n'},
		{"posix-ignores", 0, 0, 'g'},
               {"help", 0, 0, '\007'},
               {0, 0, 0, 0}
 };

  progname = argv[0];
  swlib_utilname_set(progname);
  
  if (argc < 2) { usage(progname, long_options); exit(1); }  
  
         while (1)
           {
             int option_index = 0;

             c = ugetopt_long (argc, argv, "vdbiopgl:n_",
                        long_options, &option_index);
             if (c == -1)
                 break;

             switch (c)
               {
               case 'b':
                 oForm &= ~SWPARSE_FORM_ALL;
		 oForm |= SWPARSE_FORM_INDENT;
		 break;
               case 'd':
                 swdef=new swOPTION("");
		 break;
               case 'h':
                 swdef=new swPSFi("");
		 break;
               case 'i':
                 swdef=new swINDEX("");
		 break;
               case 200:
                 swdef=new swINSTALLED("");
		 break;
	       case 'g':
                 oForm |= SWPARSE_FORM_POLICY_POSIX_IGNORES;
		 break;

               case 'o':
                 swdef=new swINFO("");
                 break;

               case 'p':
                 //swlex_definition_file = SW_PSF;
                 swdef=new swPSF("aaaa");        
                 break;
               
	       case 'v':
		 do_debug = 1;
		/*
                 swlex_debug = 1;
                 swparse_debug = 1;
		*/
                 break;
               case '?':
                 usage(progname, long_options);
                 break;
	       case 'l':
                 if (!optarg) usage(progname, long_options);
		 swparse_atlevel=atoi(optarg);
                 break;
	       case 'n':
                 oForm &= ~SWPARSE_FORM_ALL;
                 oForm |= SWPARSE_FORM_MKUP_LEN;
                 break;
	       case '_':
                 usage(progname, long_options);
                 break;
	       case ':':
                 usage(progname, long_options);
                 break;
               default:
                 usage(progname, long_options);
                 break;
               }
          }
	  if (!swdef) {
                 usage(progname, long_options);
	  }

          if (optind < argc) {
             fd = open ( argv[optind], O_RDONLY );
             if (fd <  0) exit (4);
             swlib_strncpy(swlex_filename, argv[optind], sizeof(swlex_filename));
          } else {
             fd = STDIN_FILENO;
             strcpy(swlex_filename, "stdin" );
          }

          swdef->open_parser(fd, STDOUT_FILENO);

	  if (do_debug) {
                 swlex_debug = 1;
                 swparse_debug = 1;
	  }

          ret=swdef->run_parser(swparse_atlevel, oForm);

  delete swdef;
 // close(1);
  exit (ret == 0 ? 0 : 1);
}

